### PR TITLE
Petting zoo API

### PR DESCRIPTION
**Pull request by BolunDai0216**
I created a petting zoo environment that wraps around the `MicroRTSGridModeSharedMemVecEnv`. I tested it on the petting zoo api test and it passed.

## Checklist

- [ ] I have added a `hello_world_pettingzoo.py`
    - [ ] within this file I have implemented sampling only valid actions: see #71 as an example
- [ ] I have added a functional test case in the `tests` folder to make sure `hello_world_pettingzoo.py` runs.
- [x] I have updated dependencies in poetry
- [ ] I have added documentation in the `README.md` (probably a section)

### CODE REVIEW

**Comment by BolunDai0216** [comment]
I think I also included some of the changes to the `hello_world.py` and `vec_env.py`, please ignore them.

File: gym_microrts/petting_zoo_api.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import numpy as np
+from pettingzoo import AECEnv
+from pettingzoo.test import api_test
+from pettingzoo.utils import agent_selector
+
+from gym_microrts import microrts_ai
+from gym_microrts.envs.vec_env import MicroRTSGridModeSharedMemVecEnv
+
+
+class PettingZooMicroRTSGridModeSharedMemVecEnv(AECEnv, MicroRTSGridModeSharedMemVecEnv):
```
Comment (vwxyzjn) [inline gym_microrts/petting_zoo_api.py:16]: Could you also implement the `action_mask` interface? See https://www.pettingzoo.ml/classic/checkers as an example.

File: gym_microrts/petting_zoo_api.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    #     agent_id = env.agent_name_mapping[agent]
+    #     action = actions[agent_id, :]
+    #     env.step(action)
+
+    #     if env.num_moves > 10:
+    #         break
+
+
+if __name__ == "__main__":
+    main()
```
Comment (vwxyzjn) [inline gym_microrts/petting_zoo_api.py:162]: Please move this part of the code to the root directory in a file named `hello_world_pettingzoo.py`

File: hello_world.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+from pdb import set_trace
```
Comment (vwxyzjn) [inline hello_world.py:1]: Remove this.

File: hello_world.py [hunk 4]
```diff
@@ -9,7 +11,7 @@
 
 
 envs = MicroRTSGridModeVecEnv(
-    num_selfplay_envs=0,
+    num_selfplay_envs=2,
```
Comment (vwxyzjn) [inline hello_world.py:14]: Any reason for this change?

File: hello_world.py [hunk 5]
```diff
@@ -60,10 +62,14 @@ def sample(logits):
...
             sample(action_mask[:, 18:22]),  # produce_direction parameter
             sample(action_mask[:, 22:29]),  # produce_unit_type parameter
-            sample(action_mask[:, 29 : sum(envs.action_space.nvec[1:])]),  # attack_target parameter
+            # attack_target parameter
+            sample(action_mask[:, 29: sum(envs.action_space.nvec[1:])]),
         ),
         axis=1,
     )
+
+    set_trace()
```
Comment (vwxyzjn) [inline hello_world.py:71]: Remove this.

File: gym_microrts/envs/vec_env.py [hunk 6]
```diff
@@ -1,20 +1,21 @@
 import json
 import os
 import xml.etree.ElementTree as ET
+from pdb import set_trace
```
Comment (vwxyzjn) [inline gym_microrts/envs/vec_env.py:4]: Remove this. Any debugging lines should be kept only in the local branch.

**Review by vwxyzjn — COMMENTED** [review]
Thanks for this contribution! Please run `pre-commit run --all-files` to format the code.

File: gym_microrts/petting_zoo_api.py (hunk 1 above)
Comment (BolunDai0216) [reply gym_microrts/petting_zoo_api.py:16]: Sure!

File: hello_world.py (hunk 4 above)
Comment (BolunDai0216) [reply hello_world.py:14]: No, I just wanted to see what happens if it is not zero.

File: gym_microrts/petting_zoo_api.py (hunk 1 above)
Comment (vwxyzjn) [reply gym_microrts/petting_zoo_api.py:16]: It's already implemented here so you can just apply a wrapper to it.

https://github.com/vwxyzjn/gym-microrts/blob/3d7a42f46efbd39a0b806388b8a445fbee48d00f/gym_microrts/envs/vec_env.py#L217-L221

Comment (BolunDai0216) [reply gym_microrts/petting_zoo_api.py:11]: Do you think creating a wrapper function is sufficient, or should it be directly applied in the step function and not be exposed to the user?

**Comment by vwxyzjn** [comment]
Hey @BolunDai0216, thanks for this PR. I have added a few things: 1) added pettingzoo to poetry, 2) added rendering.

Also, please run `poetry run pre-commit run --all-files` to format code and remove unused imports.

**Comment by vwxyzjn** [comment]
`bool(done[i].astype)` is basically `bool(False.astype)`, which will evaluate to true :) So I went ahead and fixed it.

**Comment by vwxyzjn** [comment]
@BolunDai0216 Fixed a few things now it kind of works:

https://user-images.githubusercontent.com/5555347/158036008-802c0802-760f-46b7-aef0-321fb45e3468.mp4

**Comment by vwxyzjn** [comment]
I have added a checklist of todo items. @BolunDai0216 if you could work on those that will be great :D Thanks again for this awesome contribution.

**Comment by BolunDai0216** [comment]
> I have added a checklist of todo items. @BolunDai0216 if you could work on those that will be great :D Thanks again for this awesome contribution.

Thanks for listing the tasks! I will try to tackle them.

**Comment by vwxyzjn** [comment]
Hey just following up with this. I think just adding a simple test case and a section in the readme would be enough

**Comment by BolunDai0216** [comment]
> Hey just following up with this. I think just adding a simple test case and a section in the readme would be enough

Sure, I plan to do it tonight.

**Comment by BolunDai0216** [comment]
> * Introduce efficient valid action sampling in `hello_world.py` #71

I see that you changed the keywork "mask" in the dict returned by observe() to "action_mask". I think given that there is a difference in the definition of what the action mask does in pettingzoo and what we currently have it will not pass the pettingzoo api test. That was the reason I named it "mask" instead. What are your thoughts on this?

**Comment by vwxyzjn** [comment]
> I see that you changed the keywork "mask" in the dict returned by observe() to "action_mask". Given that there is a difference in the definition of what the action mask does in pettingzoo and what we currently have, it will not pass the pettingzoo api test. That was the reason I named it "mask" instead. What are your thoughts on this?

Oh really? Hmm, but that chess environment (https://www.pettingzoo.ml/classic/chess) has the same observation dictionaries; why would it pass the API test cases?

**Comment by BolunDai0216** [comment]
> > I see that you changed the keywork "mask" in the dict returned by observe() to "action_mask". Given that there is a difference in the definition of what the action mask does in pettingzoo and what we currently have, it will not pass the pettingzoo api test. That was the reason I named it "mask" instead. What are your thoughts on this?
> 
> Oh really? Hmm, but that chess environment (https://www.pettingzoo.ml/classic/chess) has the same observation dictionaries; why would it pass the API test cases?

I think in the case of checkers the action mask assigns a 0 or 1 value to every possible action (e.g. move from one position to another one). In our case the action mask assigns 0 and 1s to create the legal action space.

**Comment by vwxyzjn** [comment]
> I think in the case of checkers the action mask assigns a 0 or 1 value to every possible action (e.g. move from one position to another one). In our case the action mask assigns 0 and 1s to create the legal action space.

Our action mask would also assign 0 or 1 to every possible action… I think?

**Comment by vwxyzjn** [comment]
So with the current definition the petting zoo test case would not pass?

**Comment by BolunDai0216** [comment]
> So with the current definition the petting zoo test case would not pass?

Yeah, I think so. Naming it "mask" or anything other than "action_mask" will bypass this issue.

**Comment by BolunDai0216** [comment]
I accidentally clicked request approve...

**Comment by benblack769** [comment]
So the only part of the api tests that look at the 'action_mask' part of the observation is the loop that plays through the actions:

```
        elif isinstance(prev_observe, dict) and 'action_mask' in prev_observe:
            action = random.choice(np.flatnonzero(prev_observe['action_mask']))
        else:
            action = env.action_space(agent).sample()
```

**Comment by benblack769** [comment]
This is not meant to be a strong gate claiming that the action mask should be used in a particular way, but it was just useful for some of the environments currently in pettingzoo. We can certainly discuss changes to this loop to make it work for your case.

**Comment by benblack769** [comment]
@vwxyzjn I've definitely considered having the action space contain the illegal action logic, however, it does not seem like the right abstraction. The action space is some static meta-information about the action, allowing for creation of persistent functions like neural networks. It does not appeal to me that the action space would be overloaded to also contain this state-dependent action masking logic.

However, I do wonder if action sampling/masking is worth having better builtin support for, as openspiel or some other marl libraries have. 

One possible idea I just came up with would be `action_space.sample(obs['action_mask'])`. I.e. have the space interpret and sample from the given mask. 

Another possible idea is to to simply have a utility `pettingzoo.sample(action_space, obs)` which handles a bunch of different cases, including discrete and multidiscrete.

**Comment by vwxyzjn** [comment]
Hey, sorry for the late response. It has been really hectic recently. Building better validation petting zoo APIs would be a separate issue, and let's not block this PR. @BolunDai0216, sorry for the delay. I have renamed the `action_mask` key to the `action_masks` key temporarily until the pettingzoo upstream has a better API for this @benblack769 @jkterry1. 

@BolunDai0216 thanks for contributing this PR :)
